liblwgeom/lwkmeans.c

```diff
@@ -1,77 +1,195 @@
+ */
+#define KMEANS_NULL_CLUSTER -1
+
+/*
+ * If the algorithm doesn't converge within this number of iterations,
+ * it will return with a failure error code.
+ */
+#define KMEANS_MAX_ITERATIONS 1000
+
+typedef struct kmeans_config
+{
+	/* An array of objects to be analyzed. User allocates this array */
+	/* and is responsible for freeing it. */
+	/* For objects that are not capable of participating in the distance */
+	/* calculations, but for which you still want included in the process */
+	/* (for examples, database nulls, or geometry empties) use a NULL */
+	/* value in this list. All NULL values will be returned in the */
+	/* KMEANS_NULL_CLUSTER. */
+	POINT2D** objs;
+
+	/* Number of objects in the preceding array */
+	size_t num_objs;
+
+	/* An array of initial centers for the algorithm */
+	/* Can be randomly assigned, or using proportions, */
+	/* unfortunately the algorithm is sensitive to starting */
+	/* points, so using a "better" set of starting points */
+	/* might be wise. User allocates and is responsible for freeing. */
+	POINT2D** centers;
+
+	/* Number of means we are calculating, length of preceding array */
+	unsigned int k;
+
+	/* Array to fill in with cluster numbers. User allocates and frees. */
+	int* clusters;
+
+} kmeans_config;
+
+static void
+update_r(kmeans_config* config)
 {
-	POINT2D *pa = (POINT2D*)a;
-	POINT2D *pb = (POINT2D*)b;
+	unsigned int i;
+	assert(config->objs != NULL);
+	assert(config->num_objs > 0);
+	assert(config->centers);
+	assert(config->clusters);
+	for (i = 0; i < config->num_objs; i++)
+	{
+		double distance, curr_distance;
+		int cluster, curr_cluster;
+		POINT2D* obj;
+
+		obj = config->objs[i];
+
+		/*
+		 * Don't try to cluster NULL objects, just add them
+		 * to the "unclusterable cluster"
+		 */
+		if (!obj)
+		{
+			config->clusters[i] = KMEANS_NULL_CLUSTER;
+			continue;
+		}
 
-	double dx = (pa->x - pb->x);
-	double dy = (pa->y - pb->y);
+		/* Initialize with distance to first cluster */
+		curr_distance = distance2d_sqr_pt_pt(obj, config->centers[0]);
+		curr_cluster = 0;
 
-	return dx*dx + dy*dy;
+		/* Check all other cluster centers and find the nearest */
+		for (cluster = 1; cluster < config->k; cluster++)
+		{
+			distance = distance2d_sqr_pt_pt(obj, config->centers[cluster]);
+			if (distance < curr_distance)
+			{
+				curr_distance = distance;
+				curr_cluster = cluster;
+			}
+		}
+
+		/* Store the nearest cluster this object is in */
+		config->clusters[i] = curr_cluster;
+	}
 }
 
-static void lwkmeans_pt_centroid(const Pointer * objs, const int * clusters, size_t num_objs, int cluster, Pointer centroid)
+static void
+update_means(kmeans_config* config)
 {
-	int i;
-	int num_cluster = 0;
-	POINT2D sum;
-	POINT2D **pts = (POINT2D**)objs;
-	POINT2D *center = (POINT2D*)centroid;
-
-	sum.x = sum.y = 0.0;
+	unsigned int i;
+	unsigned int* weights;
 
-	if (num_objs <= 0) return;
+	weights = lwalloc(sizeof(int) * config->k);
```

Comment:
Consider moving this to config, so you don't have to alloc/free it on each iteration.

Reply:
Moved out of function. Got rid of config altogether, -50 lines.